batchflow/utils.py

```diff
@@ -436,23 +436,29 @@ def get_available_gpus(n=1, min_free_memory=0.9, max_processes=2):
     nvidia_smi.nvmlInit()
     n_devices = nvidia_smi.nvmlDeviceGetCount()
 
-    available_devices = []
+    available_devices, memory_usage = [], []
     for i in range(n_devices):
         handle = nvidia_smi.nvmlDeviceGetHandleByIndex(i)
         info = nvidia_smi.nvmlDeviceGetMemoryInfo(handle)
 
         fraction_free = info.free / info.total
-
         num_processes = len(nvidia_smi.nvmlDeviceGetComputeRunningProcesses(handle))
 
-        if (fraction_free > min_free_memory) & (num_processes <= max_processes):
+        consider_available = (fraction_free > min_free_memory) & (num_processes <= max_processes)
+        if consider_available:
             available_devices.append(i)
+            memory_usage.append(fraction_free)
+
+        if log:
+            print(f'Device {i} | Free memory: {fraction_free:4.2f} | '
+                  f'Number of running processes: {num_processes:>2} | Free: {consider_available}')
 
     if len(available_devices) < n:
         raise ValueError(f'Not enough free devices: requested {n}, found {len(available_devices)}')
-    return available_devices[:n]
+    available_devices = np.array(available_devices)[np.argsort(memory_usage)[::-1]]
+    return sorted(available_devices[:n])
```

Comment:
why sort again?

Reply:
The first sort returns the list of GPUs, sorted by free memory, so we take 'the most free devices'
The second sort makes this list an increasing sequence of device indices